File: source/graphics/renderable.cpp
#include "../../include/graphics/renderable.h"
#include "../../include/core/camera.h"
#include "../../include/core/component.h"
#include "../../include/core/console.h"
#include "../../include/graphics/render_pipeline.h"
#include "../../include/core/time.h"

#include <glm/glm.hpp>

Renderable::Renderable() {
	objectTransform = new Transform();
	RenderPipeline::renderables.push_back(*this);
}
Renderable::Renderable(Mesh mesh) {
	this->mesh = mesh;

	objectTransform = new Transform();
	RenderPipeline::renderables.push_back(*this);
}
Renderable::Renderable(Mesh mesh, Material material) {
	this->mesh = mesh;
	this->material = material;

	objectTransform = new Transform();
	RenderPipeline::renderables.push_back(*this);
}
void Renderable::draw(Camera camera) {
	if (mesh.count == 0)
		return;

	modelMatrix = glm::mat4_cast(objectTransform->get_rotation());
	glm::translate(modelMatrix, objectTransform->get_position());
	glm::scale(modelMatrix, objectTransform->get_scale());

	material.use();

	material.shader.set_matrix4("uModel", modelMatrix);
	material.shader.set_matrix4("uView", camera.viewMatrix);
	material.shader.set_matrix4("uProjection", camera.projectionMatrix);
	material.shader.set_matrix4("uLightSpace", RenderPipeline::sunCam.projectionMatrix * RenderPipeline::sunCam.viewMatrix);

	material.shader.set_vector("uLightPos", RenderPipeline::sunCam.transform.get_position());
	material.shader.set_vector("uViewPos", camera.transform.get_position());
	material.shader.set_float("uTime", Time::time);

	mesh.draw();
}
void Renderable::draw(Camera camera, Shader shader) {
	if (mesh.count == 0)
		return;

	modelMatrix = glm::mat4_cast(objectTransform->get_rotation());
	glm::translate(modelMatrix, objectTransform->get_position());
	glm::scale(modelMatrix, objectTransform->get_scale());

	shader.use();

	shader.set_matrix4("uModel", modelMatrix);
	shader.set_matrix4("uView", camera.viewMatrix);
	shader.set_matrix4("uProjection", camera.projectionMatrix);
	shader.set_matrix4("uLightSpace", RenderPipeline::sunCam.projectionMatrix * RenderPipeline::sunCam.viewMatrix);

	shader.set_vector("uLightPos", RenderPipeline::sunCam.transform.get_position());
	shader.set_vector("uViewPos", camera.transform.get_position());
	shader.set_float("uTime", Time::time);

	/*glUniformMatrix4fv(shader.uniformLocations.at("uModel"), 1, GL_FALSE, glm::value_ptr(modelMatrix));
	glUniformMatrix4fv(shader.uniformLocations.at("uView"), 1, GL_FALSE, glm::value_ptr(camera.viewMatrix));
	glUniformMatrix4fv(shader.uniformLocations.at("uProjection"), 1, GL_FALSE, glm::value_ptr(camera.projectionMatrix));
	glUniformMatrix4fv(shader.uniformLocations.at("uLightSpace"), 1, GL_FALSE,
		glm::value_ptr(RenderPipeline::sunCam.projectionMatrix * RenderPipeline::sunCam.viewMatrix));*/

	mesh.draw();
}
void Renderable::destroy() {
	mesh.destroy();
	material.destroy();
}

File: CSharp/ASE_CSharp/Graphics/Renderable.cs
using OpenTK.Graphics.OpenGL4;
using OpenTK.Mathematics;

namespace ASE.Graphics
{
    public struct Renderable
    {
        private Transform transform;
        public Mesh mesh;
        public Material material;
        public Matrix4 modelMatrix;

        public Renderable(Transform transform, Mesh mesh)
        {
            this.transform = transform;
            this.mesh = mesh;

            RenderPipeline.AddRenderable(this);
        }
        public Renderable(Transform transform, Mesh mesh, Material material)
        {
            this.transform = transform;
            this.mesh = mesh;
            this.material = material;

            RenderPipeline.AddRenderable(this);
        }
        public void Draw(Camera camera)
        {
            if (mesh.count == 0 || material == null)
                return;

            Matrix4.CreateFromQuaternion(transform.rotation, out modelMatrix);
            modelMatrix *= Matrix4.CreateTranslation(transform.position);
            modelMatrix *= Matrix4.CreateScale(transform.scale);

            material.Use();

            material.shader.SetMatrix4("uModel", modelMatrix);
            material.shader.SetMatrix4("uView", camera.viewMatrix);
            material.shader.SetMatrix4("uProjection", camera.projectionMatrix);
            material.shader.SetMatrix4("uLightProjection", RenderPipeline.sunCam.projectionMatrix);
            material.shader.SetMatrix4("uLightView", RenderPipeline.sunCam.viewMatrix);

            GL.ActiveTexture(TextureUnit.Texture0);
            GL.BindTexture(TextureTarget.Texture2D, RenderPipeline.shadowTex);
            material.shader.SetTexture("uShadowTex0", 0);

            material.shader.SetVector("uLightPos", RenderPipeline.sunCam.transform.position);
            material.shader.SetVector("uViewPos", camera.transform.position);
            material.shader.SetFloat("uTime", Time.time);

            mesh.Draw();
        }
        public void Draw(Camera camera, Shader shader)
        {
            if (mesh.count == 0)
                return;

            Matrix4.CreateFromQuaternion(transform.rotation, out modelMatrix);
            modelMatrix *= Matrix4.CreateTranslation(transform.position);
            modelMatrix *= Matrix4.CreateScale(transform.scale);

            shader.Use();

            shader.SetMatrix4("uModel", modelMatrix);
            shader.SetMatrix4("uView", camera.viewMatrix);
            shader.SetMatrix4("uProjection", camera.projectionMatrix);
            //shader.SetMatrix4("uLightSpace", RenderPipeline.sunCam.projectionMatrix * RenderPipeline.sunCam.viewMatrix);

            //shader.SetVector("uLightPos", RenderPipeline.sunCam.transform.position);
            //shader.SetVector("uViewPos", camera.transform.position);
            shader.SetFloat("uTime", Time.time);

            mesh.Draw();
        }
    }
}
